ARROW-15347: [Doc][Guide] Update testing section in new contributors guide (#12375)
apache/arrow: docs/source/developers/guide/step_by_step/testing.rst

.. tabs::

   .. tab:: PyArrow

      We use `pytest <https://docs.pytest.org/en/latest/>`_ for
      unit tests in Python. For more info about the required
      packages see
      :ref:`Python unit testing section <python-unit-testing>`.

      **Structure**

      Test layout in PyArrow follows ``pytest`` structure for
      `Tests as part of application code <https://docs.pytest.org/en/6.2.x/goodpractices.html#tests-as-part-of-application-code>`_:

>       .. code:: console

Reviewer: ```suggestion
      .. code::
```

         pyarrow/
             __init__.py
             csv.py
             dataset.py
             ...
             tests/
                 __init__.py
                 test_csv.py
                 test_dataset.py
                 ...

      Tests for Parquet are located in a separate folder ``pyarrow/tests/parquet/``.

      .. Note::

         Good to read:
         `Invoking pytest versus python -m pytest <https://docs.pytest.org/en/6.2.x/pythonpath.html#pytest-vs-python-m-pytest>`_

>          (bottom of the page).

Reviewer: I am wondering if we should give here a takeaway for our contributors? Or is there not a simple one? (I personally always use `pytest ..`, and not `python -m pytest` out of habit, and didn't run into problems with my current setup)

Author: Yeah, I should change `python -m pytest` to `pytest...`, also to match https://arrow.apache.org/docs/developers/python.html#unit-testing

As for the takeaway, maybe instead of linking I could add this as a note:
"You can also run the tests with `python -m pytest [...]`which is almost equivalent to using `pytest [...]` directly, except that calling via python will also add the current directory to `sys.path` and can in some cases help if `pytest [...]` doesn't seem to work."

What do you think?

         $ python -m pytest pyarrow

      **Recompiling PyArrow or Arrow C++**

      If the tests start failing, try to recompile PyArrow or
      Arrow C++. See note in the :ref:`build_libraries_guide`
      section under the PyArrow tab.

      **Fixtures**

>       Inside pPyArrow tests files there can be helper functions

Reviewer: ```suggestion
      Inside PyArrow test files there can be helper functions
```

      and fixtures defined. Also other pytest decorators such as
      ``@parametrize`` or ``@skipif`` are used.

      For example:

      * ``_alltypes_example`` in ``test_pandas`` supplies a
        dataframe with 100 rows for all data types.
      * ``_check_pandas_roundtrip`` in ``test_pandas`` asserts if the
        roundtrip from ``Pandas`` through ``pa.Table`` or
        ``pa.RecordBatch`` back to ``Pandas`` yields the same result.
      * ``large_buffer`` fixture supplying a PyArrow buffer of fixed
        size to the function ``test_primitive_serialization(large_buffer)``
        in ``test_serialization.py``.

      For this reason it is good to look through the file you
      are planning to add the tests to and see if any of
      the defined functions or fixtures will be helpful.

      For more information about ``pytest`` in general visit

>       `Full pytest documentation <https://docs.pytest.org/en/6.2.x/contents.html>`_

Reviewer: ```suggestion
      `Full pytest documentation <https://docs.pytest.org/en/stable/contents.html>`_
```

(then the link doesn't get outdated)
